astropy/table/table.py

```diff
@@ -2021,6 +2021,70 @@ def group_by(self, keys):
 
         return groups.table_group_by(self, keys)
 
+    def to_pandas(self):
+        """
+        Return a :class:`pandas.DataFrame` instance
+
+        Returns
+        -------
+        dataframe : :class:`pandas.DataFrame`
+            A pandas :class:`pandas.DataFrame` instance
+        """
+        from pandas import DataFrame
+
+        if self.has_mixin_columns:
+            raise ValueError("Cannot convert a table with mixin columns to a pandas DataFrame")
+
+        for column in self.columns:
+            if len(self.dtype[column].shape) > 0:
+                raise ValueError("Cannot convert a table with multi-dimensional columns to a pandas DataFrame")
+
+        if self.masked:
+            # We have to add the columns manually so as to treat the mask
+            # correctly
+            d = DataFrame()
+            for name, column in self.columns.items():
+                if isinstance(column, MaskedColumn):
+                    if column.dtype.kind in ['i']:
+                        d[name] = column.astype(float).filled(np.nan)
+                    elif column.dtype.kind in ['f', 'c']:
+                        d[name] = column.filled(np.nan)
+                    else:
+                        d[name] = column.astype(np.object).filled(np.nan)
```

Comment:
:+1: to going via an `OrderDict` rather than `as_array`. I like that in that case `DataFrame` gets set in only one place. Furthermore, it would also still allow the check on column dimensionality to be done in an `try/except` clause -- I mean to *keep* the if statement, but only execute it if there was an actual error:
```
try:
    d = DataFrame(out)
except:
    if any(...):
       raise ValueError("Multi...")
    else:
       raise
```
The advantage (I think) is that the normal path is both as readable and as fast as it can possibly be, while the exception is still informative.